packages/app/src/elements/audit.ts

```diff
@@ -0,0 +1,239 @@
+                            weakPasswordItemIds.add(item.id);
+                        }
+
+                        // TODO: Save audit match boolean in field
+                        // passwordField.auditResult = { ...(passwordField.auditResult || {}), isWeak };
+                    }
+
+                    // TODO: Perform compromised audit
+                }
+
+                // TODO: Save audit match booleans in item
+                // item.auditResult = { lastAudited: new Date(), isReused, isWeak, isCompromised };
+
+                console.log({ item, isReused, isWeak });
+            }
+        }
+
+        // TODO: Sync?
+
+        // This makes the UI update
+        this._reusedPasswords = reusedPasswords;
+        this._weakPasswords = weakPasswords;
+        this._compromisedPasswords = compromisedPasswords;
+
+        this._countElements.forEach((countElement) => {
+            animateElement(countElement, { animation: "bounce" });
+        });
+    }
+
+    private _renderSection(listItems: ListItem[], label: string, routeParams: any) {
+        return html`
+            <section class="box count" ?hidden=${listItems.length === 0}>
+                <h2 class="uppercase bg-dark border-bottom semibold center-aligning spacing horizontal layout">
+                    <div></div>
+                    <div>${label}</div>
+                    <div class="tiny bold tag">${listItems.length}</div>
+                    <div class="stretch"></div>
+                </h2>
+                <pl-list>
+                    ${listItems.slice(0, 5).map(
+                        (listItem) => html`
+                            <div class="list-item hover click" @click=${() => this.go(`items/${listItem.item.id}`)}>
+                                <pl-vault-item-list-item
+                                    .item=${listItem.item}
+                                    .vault=${listItem.vault}
+                                    class="stretch collapse"
+                                ></pl-vault-item-list-item>
+                            </div>
+                        `
+                    )}
+                </pl-list>
+                <pl-button
+                    class="slim margined transparent"
+                    @click=${() => this.go("items", routeParams)}
+                    ?hidden=${listItems.length < 6}
+                >
+                    <div>${$l("Show All")}</div>
+                    <pl-icon icon="arrow-right" class="small left-margined"></pl-icon>
+                </pl-button>
+            </section>
+        `;
+    }
+}
+
+async function sha1(stringToHash: string) {
+    const hashedPasswordData = await crypto.subtle.digest("SHA-1", new TextEncoder().encode(stringToHash));
+
+    const hashedPassword = Array.from(new Uint8Array(hashedPasswordData))
+        .map((byte) => byte.toString(16).padStart(2, "0"))
+        .join("");
+
+    return hashedPassword;
+}
+
+function isPasswordWeak(password: string) {
```

Comment:
Let me know if there are any obvious "this password is weak" checks I can do here that I missed.

Reply:
Got [something](https://github.com/padloc/padloc/blob/master/packages/app/src/lib/util.ts#L25) for that as well 😊

```ts
const { score } = await passwordStrength(password);
return score < 2;
```

This uses [zxcvbn](https://github.com/dropbox/zxcvbn), which to this day is the most reliable tool for guessing password entropy (or "strength") that I know of. If you want you can do a little research into whether there is something better out there by now but it's served me pretty well so far!